lib/semantics/check-declarations.cc

```diff
@@ -88,31 +88,44 @@ void CheckHelper::Check(Symbol &symbol) {
   }
   if (IsAssumedLengthCharacterFunction(symbol)) {  // C723
     if (symbol.attrs().test(Attr::RECURSIVE)) {
-      context_.Say(
+      messages_.Say(
           "An assumed-length CHARACTER(*) function cannot be RECURSIVE"_err_en_US);
     }
     if (symbol.Rank() > 0) {
-      context_.Say(
+      messages_.Say(
           "An assumed-length CHARACTER(*) function cannot return an array"_err_en_US);
     }
     if (symbol.attrs().test(Attr::PURE)) {
-      context_.Say(
+      messages_.Say(
           "An assumed-length CHARACTER(*) function cannot be PURE"_err_en_US);
     }
     if (symbol.attrs().test(Attr::ELEMENTAL)) {
-      context_.Say(
+      messages_.Say(
           "An assumed-length CHARACTER(*) function cannot be ELEMENTAL"_err_en_US);
     }
     if (const Symbol * result{FindFunctionResult(symbol)}) {
       if (result->attrs().test(Attr::POINTER)) {
-        context_.Say(
+        messages_.Say(
             "An assumed-length CHARACTER(*) function cannot return a POINTER"_err_en_US);
       }
     }
   }
   if (auto *object{symbol.detailsIf<ObjectEntityDetails>()}) {
     Check(object->shape());
     Check(object->coshape());
+    if (object->isDummy() && symbol.attrs().test(Attr::INTENT_OUT)) {
+      if (FindUltimateComponent(symbol,
+              std::function<bool(const Symbol &)>{[](const Symbol &symbol) {
+                return IsCoarray(symbol) && IsAllocatable(symbol);
+              }})) {  // C846
```

Comment:
Why not just: `if (FindUltimateComponent(symbol, [](const Symbol &symbol) {...}))`?

Reply:
Sure.